Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m

```diff
@@ -0,0 +1,133 @@
+ on using PaymentIntents for dynamic authentication.
+ */
+@interface ThreeDSPaymentIntentExampleViewController ()
+
+@end
+
+@implementation ThreeDSPaymentIntentExampleViewController
+
+- (void)pay {
+    if (![self.paymentTextField isValid]) {
+        return;
+    }
+    if (![Stripe defaultPublishableKey]) {
+        [self.delegate exampleViewController:self didFinishWithMessage:@"Please set a Stripe Publishable Key in Constants.m"];
+        return;
+    }
+    [self updateUIForPaymentInProgress:YES];
+
+    NSString *returnUrl = @"payments-example://stripe-redirect";
+    // In a more interesting app, you'll probably create your PaymentIntent as soon as you know the
+    // payment amount you wish to collect from your customer. For simplicity, this example does it once they've
+    // pushed the Pay button.
+    // https://stripe.com/docs/payments/dynamic-authentication#create-payment-intent
+    [self.delegate createBackendPaymentIntentWithAmount:@1099 returnUrl:returnUrl completion:^(STPBackendResult status, NSString *clientSecret, NSError *error) {
+        if (status == STPBackendResultFailure || clientSecret == nil) {
+            [self.delegate exampleViewController:self didFinishWithError:error];
+            return;
+        }
+
+        STPAPIClient *stripeClient = [STPAPIClient sharedClient];
+        STPPaymentIntentParams *paymentIntentParams = [[STPPaymentIntentParams alloc] initWithClientSecret:clientSecret];
+        paymentIntentParams.sourceParams = [STPSourceParams cardParamsWithCard:self.paymentTextField.cardParams];
+
+        [stripeClient confirmPaymentIntentWithParams:paymentIntentParams completion:^(STPPaymentIntent * _Nullable paymentIntent, NSError * _Nullable error) {
+            if (error) {
+                [self.delegate exampleViewController:self didFinishWithError:error];
+                return;
+            }
+
+            if (paymentIntent.status == STPPaymentIntentStatusRequiresSourceAction) {
+                self.redirectContext = [[STPRedirectContext alloc] initWithPaymentIntent:paymentIntent returnUrl:returnUrl completion:^(NSString * _Nonnull clientSecret, NSError * _Nullable error) {
+                    if (error) {
+                        [self.delegate exampleViewController:self didFinishWithError:error];
+                    }
+                    else {
+                        [stripeClient retrievePaymentIntentWithClientSecret:clientSecret completion:^(STPPaymentIntent * _Nullable paymentIntent, NSError * _Nullable error) {
+                            if (error) {
+                                [self.delegate exampleViewController:self didFinishWithError:error];
+                            } else {
+                                [self finishWithStatus:paymentIntent.status];
+                            }
+                        }];
+                    }
+                    self.redirectContext = nil; // break retain cycle
```

Comment:
so it's an intentional retain cycle? Can we call that out in like L72?

Reply:
Is there a not retain cycle way to accomplish this? Doesn't feel like we should be demonstrating to users that a retain cycle is the "right" way to do something

Reply:
I think it starts on L79 - `self` needs to retain the `redirectContext` (doesn't work otherwise), and none of the examples are using weak self, so the `redirectContext` retains `self`.

`SofortExampleViewController` and `ThreeDSExampleViewController` do the same thing, but without this note.

It's possible that it's also necessary so that `STPRedirectContext.dismissPresentedViewController` is called, and it might be intentional that we're releasing the memory early.

IMO our example code does not do a great job demonstrating best practices, instead preferring to do things compactly with very little indirection, so that "all the pieces" are in one place.

I think there are improvements that could be made, and I think there's value in keeping the examples in sync.

Reply:
makes sense. Thanks